### ISSUE TITLE
semialphabet_any

### ISSUE THREAD

**Issue by h-2** [seqan/seqan3#878]
The idea is to a have a semialphabet with fixed size that type-erases all other (semi)-alphabets of the that size; i.e. it is constructible from any other semialphabet of the same size and just copies the rank. It shall also be convertible to any other semialphabet of the same size (again just copying the rank).

The advantage is that we don't need to instantiate all of lambda twice, once for `aa10murphy` and once for `aa10li`, but that we instantiate the codepaths once for `semialphabet<10>` and then we assign-through based on a run-time decision, e.g.

```cpp
if (reduction == AlphabetEnum::MURPHY10)
    reducedSeqs = unreducedSeqs | view::convert<aa10murphy> | view::convert<semialphabet_any<10>>;
else if (reduction == AlphabetEnum::LI10)
    reducedSeqs = unreducedSeqs | view::convert<aa10li> | view::convert<semialphabet_any<10>>;
```
(the important part here that reducedSeqs has the same type even though we can assign different alphabets).

Here's a rough sketch
```cpp
template <size_t size>
class semialphabet_any : public alphabet_base<semialphabet_any<size>, size, void>
{
public:    

    template <Semialphabet other_alph_t>
    semialphabet_any(other_alph_t const other)
    {
         static_assert(alphabet_size_v<other_alph_t> == size, "fail");
         assign_rank(to_rank(other));
    }

    template <Semialphabet other_alph_t>
        requires alphabet_size_v<other_alph_t> == size
    operator other_alph_t()
    {
         other_alph_t other{};
         assign_rank(other, to_rank());
         return other;
    }
};
```

**Comment by rrahn** [comment]
@h-2 This is very cool but also very limited. I am wondering if this should be part of the library then or not something that is only relevant to lambda? I mean I cannot type-erase dna4, dna5 or aa27 with this. It would be cool if internally all alphabets until a maximum size of 256 could be represented and it provides internal variables for the assigned size etc.. But I think this won't work unless we drop the requirement that alphabet_size is a compile time constant. Or the alphabet size is always 256. I mean in both cases you cannot put the alphabet in a bitcompressed vector. I am not sure where else we need the size information of the alphabet to be available at compile time. If not we could make the alphabet size also be a runtime information. And places that require alphabet_size_v then need to explicitly constraint this. So the former suggestion would allow to ask for the alphabet size, but not guaranteed at compile time.
Note, this is just some brain storming, but I think it should be discussed.

**Comment by h-2** [comment]
I have a plan for a different `alphabet_any<char_type>` that type erases through the char type (and has an alphabet size of the size of the char type). Does that address your concerns?

I am not convinced we should drop the fixed size, a lot of stuff depends on it: bitcompression, conversion tables, scoring matrixes and so on. We wouldn't even be able to pre-select types for our alphabet rank type a.s.o.

**Comment by rrahn** [comment]
> I have a plan for a different alphabet_any<char_type> that type erases through the char type (and has an alphabet size of the size of the char type). Does that address your concerns?

This is what I meant. 

> I am not convinced we should drop the fixed size, a lot of stuff depends on it: bitcompression, conversion tables, scoring matrixes and so on. We wouldn't even be able to pre-select types for our alphabet rank type a.s.o.

Yes, I see this as well as a problem. But somewhat the char based erasure would be kind the thing that makes it more generic.

**Comment by sarahet** [comment]
Implemented by #981
